contracts/delegation/dao-vote-delegation/src/contract.rs

```diff
@@ -0,0 +1,807 @@
+use cosmwasm_std::{ensure, Addr, Order, Uint128};
+#[cfg(not(feature = "library"))]
+use cosmwasm_std::{
+    entry_point, to_json_binary, Binary, Decimal, Deps, DepsMut, Env, MessageInfo, Response,
+    StdResult,
+};
+use cw2::{get_contract_version, set_contract_version};
+use cw_paginate_storage::paginate_map_keys;
+use cw_storage_plus::Bound;
+use cw_utils::{maybe_addr, nonpayable};
+use dao_interface::helpers::OptionalUpdate;
+use dao_interface::state::{ProposalModule, ProposalModuleStatus};
+use dao_interface::voting::InfoResponse;
+use dao_voting::delegation::{
+    calculate_delegated_vp, DelegationResponse, RegistrationResponse,
+    UnvotedDelegatedVotingPowerResponse, VotingPowerCapResponse,
+};
+use dao_voting::voting;
+use semver::Version;
+
+use crate::helpers::{
+    add_delegated_vp, ensure_max_delegations_not_reached, ensure_setup, get_udvp, get_voting_power,
+    is_delegate_registered, remove_delegated_vp, unregister_delegate,
+};
+use crate::hooks::{
+    execute_membership_changed, execute_nft_stake_changed, execute_stake_changed, execute_vote_hook,
+};
+use crate::msg::{
+    DelegateResponse, DelegatesResponse, DelegationsResponse, ExecuteMsg, InstantiateMsg,
+    MigrateMsg, QueryMsg,
+};
+use crate::state::{
+    Config, Delegate, Delegation, CONFIG, DAO, DELEGATED_VP, DELEGATES, DELEGATIONS,
+    DELEGATION_ENTRIES, PERCENT_DELEGATED, PROPOSAL_HOOK_CALLERS, VOTING_POWER_HOOK_CALLERS,
+    VP_CAP_PERCENT,
+};
+use crate::ContractError;
+
+pub(crate) const CONTRACT_NAME: &str = "crates.io:dao-vote-delegation";
+pub(crate) const CONTRACT_VERSION: &str = env!("CARGO_PKG_VERSION");
+
+pub const DEFAULT_LIMIT: u32 = 10;
+
+/// in tests on Neutron, with a block max gas of 30M (which is one of the lowest
+/// gas limits on any chain), we found that 50 delegations is a safe upper
+/// bound, so this defaults to 50.
+pub const DEFAULT_MAX_DELEGATIONS: u64 = 50;
```

Comment:
honestly, while technically this is a limitation, I like the way it works.

this places the gas burden throughout the delegation (and proposal) lifecycle almost entirely on the delegators. that feels right - the delegates should not have to worry about stuff like monitoring their delegators which may be misbehaving or something.